### PR TITLE
Migrate from avro_rs (0.13) to apache_avro (0.14)

### PR DESCRIPTION
**Pull request by martin-g**
Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2783.

 # Rationale for this change

avro_rs has been [donated](https://github.com/flavray/avro-rs/#%EF%B8%8F-avro-rs-is-no-longer-maintained-here-please-refer-to-the-apacheavrorust-repository-instead-make-sure-you-read-the-apache-avro-guidelines-here-%EF%B8%8F) to Apache Avro project.

# What changes are included in this PR?

# Are there any user-facing changes?
Yes, some APIs have changed.
Now is the moment to discuss them! But since it will be `0.x` release we could make more breaking changes later too!

### CODE REVIEW

File: datafusion/common/Cargo.toml [hunk 1]
```diff
@@ -33,13 +33,14 @@ name = "datafusion_common"
...
 
 [features]
-avro = ["avro-rs"]
+avro = ["apache-avro"]
 jit = ["cranelift-module"]
 pyarrow = ["pyo3"]
 
 [dependencies]
+cranelift-module = { version = "0.84.0", optional = true }
+apache-avro = { path = "/home/martin/git/apache/avro/lang/rust/avro", features = ["snappy"], optional = true }
```
Comment (martin-g) [inline datafusion/common/Cargo.toml:42]: This is temporary until apache-avro 0.14 is released to https://crates.io (hopefully soon!)

**Comment by martin-g** [comment]
https://crates.io/crates/apache-avro/0.14.0 has been released!

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @martin-g!

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @martin-g !

I also ran the tests manaully:
```shell
cargo test --features=avro -p datafusion
...
test src/avro_to_arrow/reader.rs - avro_to_arrow::reader::ReaderBuilder::new (line 58) ... ok
```

And the example

```shell
cargo run --features=datafusion/avro -p datafusion-examples --example avro_sql
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `/Users/alamb/Software/target-df/debug/examples/avro_sql`
+---------+------------+----------------------------------------------+
| int_col | double_col | CAST(alltypes_plain.date_string_col AS Utf8) |
+---------+------------+----------------------------------------------+
| 1       | 10.1       | 03/01/09                                     |
| 1       | 10.1       | 04/01/09                                     |
| 1       | 10.1       | 02/01/09                                     |
+---------+------------+----------------------------------------------+
```

cc @Igosuki  who contributed the initial avro implementation

**Pull request merged**